File: C# DB/Databases Advanced - Entity Framework/06 C# OOP Intro Ex/OOP Introduction - Exerciseс/07. Animals/Animal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

class Animal
{
    private string name;
    private int age;
    private string gender;

    public Animal()
    {

    }

    public Animal(string name, int age)
    {
        this.Name = name;
        this.Age = age;
    }

    public Animal(string name, int age, string gender)
    {
        this.Name = name;
        this.Age = age;
        this.Gender = gender;
    }

    public string Name
    {
        get
        {
            return this.name;
        }
        set
        {
            if(string.IsNullOrEmpty(value))
            {
                throw new ArgumentException("Invalid input!");
            }
            this.name = value;
        }
    }

    public int Age
    {
        get
        {
            return this.age;
        }
        set
        {
            if (value < 0)
            {
                throw new ArgumentException("Invalid input!");
            }
            this.age = value;
        }
    }
    public string Gender
    {
        get
        {
            return this.gender;
        }
        set
        {
            if (string.IsNullOrEmpty(value) || (value.ToLower() != "male" && value.ToLower() != "female"))
            {
                throw new ArgumentException("Invalid input!");
            }
            this.gender = value;
        }
    }

    public virtual string ProduceSound()
    {
        return String.Empty;
    }

    public override string ToString()
    {
        return $"{this.GetType().Name}" + Environment.NewLine +
            $"{this.Name} {this.Age} {this.Gender}" + Environment.NewLine + $"{this.ProduceSound()}";
    }
}

File: Java/01.Advanced/02.OOP/05.Inheritance/Ex/src/t_06_animals/animals/Animal.java
package t_06_animals.animals;

public abstract class Animal {
    private final String name;
    private final int age;
    private final String gender;

    protected Animal(String name, int age, String gender) {
        this.name = name;
        this.age = age;
        this.gender = gender;
    }

    protected abstract String produceSound();

    public String getName() {
        return name;
    }

    public int getAge() {
        return age;
    }

    public String getGender() {
        return gender;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(26);

        sb.append(this.getClass().getSimpleName());
        sb.append("\r");
        sb.append("\n");

        sb.append(name);
        sb.append(' ');
        sb.append(age);
        sb.append(' ');
        sb.append(gender);
        sb.append("\r");
        sb.append("\n");

        sb.append(this.produceSound());

        return sb.toString();
    }
}
